app/models/fact_value.rb

```diff
@@ -81,13 +81,16 @@ def self.count_each(fact, options = {})
     output
   end
 
-  def self.build_facts_hash(facts)
-    hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
-
+  def self.build_facts_hash(facts, has_host = false)
     hash = {}
+    hosts = Host.where(:id => facts.group_by(&:host_id).keys).all  if has_host
     facts.each do |fact|
+      hosts = Host.where(:id => facts.group_by(&:host_id).keys).all unless has_host
```

Comment:
This line should be outside of each, otherwise it will be executed once per each fact.
I can see that the condition is unnecessary too - you are performing the same query whether `has_host` is true or not.